src/collapse/perf.rs

```diff
@@ -324,6 +387,91 @@ impl PerfState {
         }
         Ok(())
     }
+
+    // Use addr2line to determine the symbol to use for each program counter address
+    // (as opposed to using the symbol names that perf script produces)
+    fn un_inline(&mut self, addr: &str, module: &str) -> result::Result<String, Error> {
+        let addr = u64::from_str_radix(addr, 16).map_err(|e| {
+            Error::Parse(format!(
+                "unable to parse {} as hex address, caused by: {}",
+                addr, e
+            ))
+        })?;
+
+        if let Entry::Vacant(entry) = self.addr2line_contexts.entry(module.into()) {
+            let bytes = fs::read(module)?;
+            let file = &object::File::parse(&bytes).map_err(|e| {
+                Error::Parse(format!(
+                    "unable to parse file: {}, caused by: {}",
+                    module, e
+                ))
+            })?;
+            entry.insert(addr2line::Context::new(file)?);
+        };
+        let ctx = &self.addr2line_contexts[module];
+
+        let mut frames = ctx.find_frames(addr).map_err(|e| {
+            Error::Parse(format!(
+                "unable to parse frames from module: {}, at address: {}. caused by: {}",
+                module, addr, e
+            ))
+        })?;
+
+        let mut full_func = SmallVec::<[String; 1]>::new();
+        let mut total_length = 0;
+        while let Some(frame) = frames.next().unwrap() {
+            let mut function = frame
+                .function
+                .map(|func| {
+                    let name = remove_discriminator(
+                        func.raw_name()
+                            .map(Cow::from)
+                            .unwrap_or_else(|_| Cow::from("??")),
+                    );
+                    strip_rust_hash(addr2line::demangle_auto(name, func.language).as_ref())
+                        .to_string()
+                })
+                .unwrap_or_else(|| "??".into());
+
+            if self.opt.show_context {
```

Comment:
The Perl code does some really funky stuff [around here](https://github.com/brendangregg/FlameGraph/blob/f857ebc94bfe2a9bfdc4f1536ebacfb7466f69ba/stackcollapse-perf.pl#L140L144) when `--context` is enabled. Any idea how that fits into here?

Reply:
When the -f option for addr2line is used the output alternates as follows:

FUNCTIONNAME
FILENAME:LINENO
FUNCTIONNAME
FILENAME:LINENO

When --context is enabled, the perl script uses both the FUNCTIONNAME line and the FILENAME:LINENO line. It saves the function name line in $one_item and gets the location info during the next iteration of the loop. When --context is disabled it only cares about every other line.

I believe my Rust version matches this behavior, but it loops on frames instead of lines. If --context is enabled it tries to include the filename and line number from frame.location. If --context is disabled it only includes the function name.

Does that make sense?

Reply:
Ah, I see, so we replicate this behavior through the `:`-separated `format!` we do to produce each name of the trace. That seems reasonable.